hive: Unwrap provider inputs and outputs in PrintObjects

joamaki opened the pull request:
Before the config objects shown in PrintObjects showed the defaults by iterating over the registered flags:

```
    $ ./cilium-agent objects --k8s-api-server foobar

    ⚙️ client.Config:
      • enable-k8s-api-discovery[bool]  = false
      • k8s-api-server[string]          =
      • k8s-client-burst[int]           = 0
      • k8s-client-qps[float32]         = 0
      • k8s-heartbeat-timeout[duration] = 30s
      • k8s-kubeconfig-path[string]     =
```

Above you can see that the "k8s-api-server" flag did not reflect what was provided on command-line

The first commit changes the config info to pull the parsed configuration from the dig container and show that using the spew package:

```
    ⚙️ (client.Config) {
      K8sAPIServer: (string) (len=6) "foobar",
      K8sKubeConfigPath: (string) "",
      K8sClientQPS: (float32) 0,
      K8sClientBurst: (int) 0,
      K8sHeartbeatTimeout: (time.Duration) 30s,
      EnableK8sAPIDiscovery: (bool) false
    }
```


The second commit improves PrintObjects() by using dig.FillProviderInfo. Previously the provider function type was shown in PrintObjects:

```
  🚧 pkg/node.newLocalNodeStore (local_node_store.go:71): func(node.LocalNodeStoreParams) (node.LocalNodeStore, error)
```

This does not unwrap the param and result structs and thus does not give enough information about the constructor. This is now fixed by using dig.FillProviderInfo to pull the constructors actual inputs and outputs:

```
  🚧 pkg/node.newLocalNodeStore (local_node_store.go:71):
    ⮩ hive.Lifecycle, node.LocalNodeInitializer[optional]
    ⮨ node.LocalNodeStore
```

joamaki committed:
hive: Fix printing of hook function names

We need to have special handling for printing the start and stop hook
names if the hook is not the "Hook" struct.

To keep things simple just print the type name + method instead of
going via FuncForPC. This makes sure we show reasonable output for
polymorphic structs as well (which would not have file+line or the
type params if FuncForPC was used).

Log output before:
level=info msg="Start hook executed" duration="1.853µs" function="node.newLocalNodeStore.func1 (local_node_store.go:80)" subsys=hive
level=info msg="Start hook executed" duration="5.681µs" function="hive.Hook.Start-fm (<autogenerated>:1)" subsys=hive

Log output after:
level=info msg="Start hook executed" duration=992ns function="node.newLocalNodeStore.func1 (local_node_store.go:80)" subsys=hive
level=info msg="Start hook executed" duration="4.939µs" function="*resource.resource[*k8s.io/api/core/v1.Node].Start" subsys=hive

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive: Unwrap provider inputs and outputs in PrintObjects

Previously the provider function type was shown in PrintObjects:


  🚧 pkg/node.newLocalNodeStore (local_node_store.go:71): func(node.LocalNodeStoreParams) (node.LocalNodeStore, error)

This does not unwrap the param and result structs and thus does
not give enough information about the constructor. This commit fixes
the issue by using dig.FillProviderInfo to pull the constructors
actual inputs and outputs:

  🚧 pkg/node.newLocalNodeStore (local_node_store.go:71):
    ⮩ hive.Lifecycle, node.LocalNodeInitializer[optional]
    ⮨ node.LocalNodeStore

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive: Show actual parsed config in PrintObjects()

Before the config objects shown in PrintObjects showed the defaults by iterating
over the registered flags:

    $ ./cilium-agent objects --k8s-api-server foobar

    ⚙️ client.Config:
      • enable-k8s-api-discovery[bool]  = false
      • k8s-api-server[string]          =
      • k8s-client-burst[int]           = 0
      • k8s-client-qps[float32]         = 0
      • k8s-heartbeat-timeout[duration] = 30s
      • k8s-kubeconfig-path[string]     =

Above you can see that the "k8s-api-server" flag did not reflect what was provided on
command-line

This commit changes the config info to pull the parsed configuration from the dig
container and show that using the spew package:

    ⚙️ (client.Config) {
      K8sAPIServer: (string) (len=6) "foobar",
      K8sKubeConfigPath: (string) "",
      K8sClientQPS: (float32) 0,
      K8sClientBurst: (int) 0,
      K8sHeartbeatTimeout: (time.Duration) 30s,
      EnableK8sAPIDiscovery: (bool) false
    }

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive: Clean up breaks in PrintObjects

This removes AddBreak and does the more obvious right thing and prints the
separating newline in InfoNode.Print().

Fixes the "double breaks" after nested modules.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink commented on pkg/hive/cell/info.go:
```diff
@@ -4,13 +4,22 @@
+
+	// maxLineLength is the maximum line length after which the InfoLeaf line
+	// wraps.
+	maxLineLength = 120
```
Perhaps we should consider getting the terminal width [dynamically](https://pkg.go.dev/golang.org/x/term#GetSize) instead of hard coding a wrapping length.

joamaki replied:
Agreed, will add it.

dylandreimerink commented on pkg/hive/cell/info.go:
```diff
@@ -42,12 +70,13 @@ func NewInfoNode(header string) *InfoNode {
 
-func (n *InfoNode) Add(child Info) {
-	n.children = append(n.children, child)
+func (n *InfoNode) Condensed() *InfoNode {
```
Is this right? It seams strange to me to have a method with pointer receiver which modifies itself and returns itself. Perhaps this was meant to be a value receiver so it returns a copy?

In the current form we have a strange side effect where we would print differently depending on call order. Info() then Print() or Print() then Info()

joamaki replied:
Since this was meant to be used when constructing it, I'll just remove the `Condensed()` method and just set the attribute directly.

joamaki committed:
hive: Detect terminal width for use in InfoLeaf line wrapping

Detect the width with term.GetSize() instead of using a hard-coded width
for wrapping the list of input and output objects.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink reviewed (approved):
LGTM!
